### ISSUE TITLE
Bugs with new file access infrastructure?

### ISSUE THREAD

**Issue by edwardalee** [icyphy/lingua-franca#317]
The new file access infrastructure does not seem to work with the IDE's mechanism for marking (and even clearing) errors in the IDE. 

To reproduce this, create a CTests project and do a clean in the Eclipse IDE.  Unfortunately, the instructions for creating this project used to be here: [https://github.com/icyphy/lingua-franca/wiki/Regression-Tests#adding-and-editing-tests-in-eclipse]. They are replaced now with instructions for running the tests on the command line.  @lhstrh : Can we restore those instructions and put the instructions for running tests on the command line on a different page?

My IDE has an error marker on the file `tests/C/src/lib/Test.lf`. This is a spurious error marker. I don't know where it came from. If I rebuild the file, then I get the following, which complains that a resource does not exist. But it does exist.  I get lots of other errors like this.

```
Generating code for: platform:/resource/CTests/src/lib/Test.lf
******** mode: INTEGRATED
******** source file: /Users/eal/lingua-franca-master/git/lingua-franca/test/C/src/lib/Test.lf
******** generated sources: /Users/eal/lingua-franca-master/git/lingua-franca/test/C/src-gen/lib/Test
******** generated binaries: /Users/eal/lingua-franca-master/git/lingua-franca/test/C/bin
Warning: Deleting markers in the IDE failed: org.eclipse.core.internal.resources.ResourceException: Resource '/Users/eal/lingua-franca-master/git/lingua-franca/test/C/src/lib/Test.lf' does not exist.
36578 [Worker-1: Building] ERROR org.eclipse.xtext.builder.BuilderParticipant  - Error during compilation of 'platform:/resource/CTests/src/lib/Test.lf'.
java.lang.NullPointerException
	at org.icyphy.generator.GeneratorBase.setResources(GeneratorBase.java:447)
	at org.icyphy.generator.GeneratorBase.doGenerate(GeneratorBase.java:408)
	at org.icyphy.generator.CGenerator.doGenerate(CGenerator.java:492)
	at org.icyphy.generator.LinguaFrancaGenerator.doGenerate(LinguaFrancaGenerator.java:57)
	at org.eclipse.xtext.generator.GeneratorDelegate.doGenerate(GeneratorDelegate.java:44)
	at org.eclipse.xtext.generator.GeneratorDelegate.generate(GeneratorDelegate.java:35)
	at org.eclipse.xtext.builder.BuilderParticipant.handleChangedContents(BuilderParticipant.java:597)
```

**Comment by edwardalee** [comment]
Also, probably due to the same bug, if you get a warning, then now code generation fails.  Try this example (in the IDE):
```
Generating code for: platform:/resource/CTests/src/federated/HelloWorldFederatedCPP.lf
******** mode: INTEGRATED
******** source file: /Users/eal/lingua-franca-master/git/lingua-franca/test/C/src/federated/HelloWorldFederatedCPP.lf
******** generated sources: /Users/eal/lingua-franca-master/git/lingua-franca/test/C/src-gen/federated/HelloWorldFederatedCPP
******** generated binaries: /Users/eal/lingua-franca-master/git/lingua-franca/test/C/bin
Warning: Deleting markers in the IDE failed: org.eclipse.core.internal.resources.ResourceException: Resource '/Users/eal/lingua-franca-master/git/lingua-franca/test/C/src/federated/HelloWorldFederatedCPP.lf' does not exist.
--- Looking for command g++ ... SUCCESS
--- Current working directory: file://Users/eal/lingua-franca-master/git/lingua-franca/test/C
--- Executing command: g++ src-gen/federated/HelloWorldFederatedCPP/HelloWorldFederatedCPP_a.c /Users/eal/lingua-franca-master/git/lingua-franca/test/C/src-gen/federated/HelloWorldFederatedCPP/core/platform/lf_macos_support.c -o bin/HelloWorldFederatedCPP_a -pthread -DNUMBER_OF_WORKERS=1 -Wno-write-strings
clang: warning: treating 'c' input as 'c++' when in C++ mode, this behavior is deprecated [-Wdeprecated]
clang: warning: treating 'c' input as 'c++' when in C++ mode, this behavior is deprecated [-Wdeprecated]
WARNING: /Users/eal/lingua-franca-master/git/lingua-franca/test/C/src/federated/HelloWorldFederatedCPP.lf 
clang: warning: treating 'c' input as 'c++' when in C++ mode, this behavior is deprecated [-Wdeprecated]
clang: warning: treating 'c' input as 'c++' when in C++ mode, this behavior is deprecated [-Wdeprecated]
912031 [Worker-19: Building] ERROR org.eclipse.xtext.builder.BuilderParticipant  - Error during compilation of 'platform:/resource/CTests/src/federated/HelloWorldFederatedCPP.lf'.
org.eclipse.core.internal.resources.ResourceException(/Users/eal/lingua-franca-master/git/lingua-franca/test/C/src/federated/HelloWorldFederatedCPP.lf)[368]: java.lang.Exception: Resource '/Users/eal/lingua-franca-master/git/lingua-franca/test/C/src/federated/HelloWorldFederatedCPP.lf' does not exist.
	at org.eclipse.core.internal.resources.ResourceException.provideStackTrace(ResourceException.java:42)
	at org.eclipse.core.internal.resources.ResourceException.<init>(ResourceException.java:38)
	at org.eclipse.core.internal.resources.Resource.checkExists(Resource.java:330)
	at org.eclipse.core.internal.resources.Resource.checkAccessible(Resource.java:204)
	at org.eclipse.core.internal.resources.Resource.createMarker(Resource.java:693)
	at org.icyphy.generator.GeneratorBase.report(GeneratorBase.java:1839)
	at org.icyphy.generator.GeneratorBase.report(GeneratorBase.java:1893)
	at org.icyphy.generator.GeneratorBase.reportCommandErrors(GeneratorBase.java:1622)
	at org.icyphy.generator.GeneratorBase.runCCompiler(GeneratorBase.java:888)
```

**Comment by edwardalee** [comment]
I have pushed a fix to the above errors, but now discovered that errors in imported files are not being reported correctly (this was true before my fix as well).  The reason is that the fileConfig field of GeneratorBase always points to the main source .lf file, never to the imported file, so when the code generator generates `#line` directives, the filename is always the main source .lf file, never the imported file. So the C compiler reports the error in the wrong file.

**Comment by edwardalee** [comment]
I have pushed a fix so that now #line directives correctly point to the file in which the resource is defined rather than the master file. Errors are now properly highlighted in imported files. However, there is still another problem.  When code generating a file without a main reactor, the following error occurs:
```
clang: error: cannot specify -o when generating multiple output
   files
```

**Comment by edwardalee** [comment]
I have pushed a fix so that now when you compile a .lf file with no main reactor, you get a .o file in the bin directory (I'm not sure that's where we want it, but that's where it is going now). I also fixed things so that if there is no main reactor, then code is generated even for reactors that have no instantiations.  Previously, no code was generated, so errors in your C code would not show up when compiling the file.  I believe that all these issues are now resolved.
